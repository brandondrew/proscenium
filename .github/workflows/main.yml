name: Ruby

on:
  push:
    branches:
      - master

  pull_request:

jobs:
  release:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v3

      - name: Set up Ruby
        uses: ruby/setup-ruby@v1
        with:
          ruby-version: 3.1

      - uses: denoland/setup-deno@v1.0.0
        with:
          deno-version: v1.x

      - name: Compile esbuild
        run: deno task compile:esbuild:macm

      - name: Install gem dependencies
        run: bundle install

      - name: Build native gem
        env:
          PROFILE: release
        run: bundle exec rake build

      - uses: actions/upload-artifact@v3
        with:
          name: artifacts
          path: bin/esbuild-*
