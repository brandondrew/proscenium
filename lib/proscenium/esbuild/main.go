package main

// go build -buildmode=c-shared -o main.so main.go

import "C"
import "os"
import "strings"
import "encoding/json"

// import "fmt"
import "github.com/evanw/esbuild/pkg/api"

//export transform
func transform(input *C.char) *C.char {
	inputString := C.GoString(input)
	result := api.Transform(inputString, api.TransformOptions{
		Loader: api.LoaderJS,
	})

	s := string(result.Code)

	return C.CString(s)
}

var envPlugin = api.Plugin{
	Name: "env",
	Setup: func(build api.PluginBuild) {
		// Intercept import paths called "env" so esbuild doesn't attempt
		// to map them to a file system location. Tag them with the "env-ns"
		// namespace to reserve them for this plugin.
		build.OnResolve(api.OnResolveOptions{Filter: `^env$`},
			func(args api.OnResolveArgs) (api.OnResolveResult, error) {
				return api.OnResolveResult{
					Path:      args.Path,
					Namespace: "env-ns",
				}, nil
			})

		// Load paths tagged with the "env-ns" namespace and behave as if
		// they point to a JSON file containing the environment variables.
		build.OnLoad(api.OnLoadOptions{Filter: `.*`, Namespace: "env-ns"},
			func(args api.OnLoadArgs) (api.OnLoadResult, error) {
				mappings := make(map[string]string)
				for _, item := range os.Environ() {
					if equals := strings.IndexByte(item, '='); equals != -1 {
						mappings[item[:equals]] = item[equals+1:]
					}
				}
				bytes, err := json.Marshal(mappings)
				if err != nil {
					return api.OnLoadResult{}, err
				}
				contents := string(bytes)
				return api.OnLoadResult{
					Contents: &contents,
					Loader:   api.LoaderJSON,
				}, nil
			})
	},
}

//export build
func build(entrypoint *C.char) *C.char {
	entrypointString := C.GoString(entrypoint)

	result := api.Build(api.BuildOptions{
		EntryPoints: []string{entrypointString},
		Bundle:      true,
		Plugins:     []api.Plugin{envPlugin},
		Write:       false,
	})

	outputString := string(result.OutputFiles[0].Contents)
	return C.CString(outputString)
}

func main() {
	// fmt.Println(transform())
	// fmt.Printf("%s", build("input.ts"))
}
