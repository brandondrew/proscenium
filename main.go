package main

/*
struct Result {
	int success;
	char* response;
};
*/
import "C"

import (
	"encoding/json"
	"joelmoss/proscenium/internal/builder"
	"joelmoss/proscenium/internal/resolver"
	"joelmoss/proscenium/internal/types"
	"path"
	"strings"
)

// Build the given `path` in the `root`.
//
//   - path - The path to build relative to `root`.
//   - root - The working directory.
//   - env - The environment (1 = development, 2 = test, 3 = production)
//   - importMap - Path to the import map relative to `root`.
//   - debug
//
//export build
func build(filepath *C.char, root *C.char, env C.uint, importMap *C.char, debug bool) C.struct_Result {
	types.Env = types.Environment(env)

	pathStr := C.GoString(filepath)

	result := builder.Build(builder.BuildOptions{
		Path:          pathStr,
		Root:          C.GoString(root),
		ImportMapPath: C.GoString(importMap),
		Debug:         debug,
	})

	if len(result.Errors) != 0 {
		j, err := json.Marshal(result.Errors[0])
		if err != nil {
			return C.struct_Result{C.int(0), C.CString(string(err.Error()))}
		}

		return C.struct_Result{C.int(0), C.CString(string(j))}
	}

	contents := string(result.OutputFiles[0].Contents)

	isSourceMap := strings.HasSuffix(pathStr, ".map")
	if isSourceMap {
		return C.struct_Result{C.int(1), C.CString(contents)}
	}

	sourcemapUrl := path.Base(pathStr)
	if strings.HasSuffix(result.OutputFiles[0].Path, ".css") {
		contents += "/*# sourceMappingURL=" + sourcemapUrl + ".map */"
	} else {
		contents += "//# sourceMappingURL=" + sourcemapUrl + ".map"
	}

	return C.struct_Result{C.int(1), C.CString(contents)}
}

// Resolve the given `path` relative to the `root`.
//
//   - path - The path to build relative to `root`.
//   - root - The working directory.
//   - env - The environment (1 = development, 2 = test, 3 = production)
//   - importMap - Path to the import map relative to `root`.
//
//export resolve
func resolve(path *C.char, root *C.char, env C.uint, importMap *C.char) C.struct_Result {
	types.Env = types.Environment(env)

	resolvedPath, err := resolver.Resolve(resolver.Options{
		Path:          C.GoString(path),
		Root:          C.GoString(root),
		ImportMapPath: C.GoString(importMap),
	})
	if err != nil {
		return C.struct_Result{C.int(0), C.CString(string(err.Error()))}
	}

	return C.struct_Result{C.int(1), C.CString(resolvedPath)}
}

func main() {}
