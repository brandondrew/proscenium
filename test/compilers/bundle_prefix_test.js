import { assertSnapshot } from "testing/snapshot.ts";
import { beforeEach, describe, it } from "testing/bdd.ts";
import { join } from "std/path/mod.ts";

import main from "../../lib/proscenium/compilers/esbuild.js";

const root = join(Deno.cwd(), "test", "dummy");
const lightningcssBin = join(Deno.cwd(), "bin", "lightningcss");

describe("bundle: prefix", () => {
  beforeEach(() => {
    Deno.env.set("RAILS_ENV", "test");
    Deno.env.set("PROSCENIUM_TEST", "test");
  });

  it("js import", async (t) => {
    const result = await main("lib/bundle_import/index.js", {
      root,
      lightningcssBin,
      importMap: "config/import_maps/bundled.json",
    });

    await assertSnapshot(t, new TextDecoder().decode(result));
  });

  it("css import", async (t) => {
    const result = await main("lib/bundle_import/styles.css", {
      root,
      lightningcssBin,
      cssMixinPaths: [join(root, "lib")],
      importMap: "config/import_maps/bundled.json",
    });

    await assertSnapshot(t, new TextDecoder().decode(result));
  });
});
