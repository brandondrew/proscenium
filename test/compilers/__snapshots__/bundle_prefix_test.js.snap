export const snapshot = {};

snapshot[`bundle: prefix > js import 1`] = `
'// lib/bundle_import/foo.js
var foo_default = () => console.log("bundle_import/foo.js");

// lib/foo3.js
console.log("/lib/foo3.js");

// lib/foo4.js
var foo4_default = () => console.log("/lib/foo4.js");

// lib/foo2.js
console.log("/lib/foo2.js");

// packages/mypackage/index.js
console.log("node_modules/mypackage");

// lib/bundle_import/bar.js
console.log("bundle_import/bar.js");

// lib/component.jsx
import { jsx } from "/node_modules/.pnpm/react@18.2.0/node_modules/react/jsx-runtime.js";
var Component = () => {
  return /* @__PURE__ */ jsx("div", { children: "Hello" });
};
var component_default = Component;

// url:https://esm.sh/v99/email-regex@5.0.0/es2022/email-regex.js
var e = "[^\\\\\\\\.\\\\\\\\s@:](?:[^\\\\\\\\s@:]*[^\\\\\\\\s@:\\\\\\\\.])?@[^\\\\\\\\.\\\\\\\\s@]+(?:\\\\\\\\.[^\\\\\\\\.\\\\\\\\s@]+)*";
function s({ exact: n2 } = {}) {
  return n2 ? new RegExp(\`^\${e}\$\`) : new RegExp(e, "g");
}

// url:https://esm.sh/v99/ip-regex@5.0.0/es2022/ip-regex.js
var g = "[a-fA-F\\\\\\\\d:]";
var d = (e2) => e2 && e2.includeBoundaries ? \`(?:(?<=\\\\\\\\s|^)(?=\${g})|(?<=\${g})(?=\\\\\\\\s|\$))\` : "";
var c = "(?:25[0-5]|2[0-4]\\\\\\\\d|1\\\\\\\\d\\\\\\\\d|[1-9]\\\\\\\\d|\\\\\\\\d)(?:\\\\\\\\.(?:25[0-5]|2[0-4]\\\\\\\\d|1\\\\\\\\d\\\\\\\\d|[1-9]\\\\\\\\d|\\\\\\\\d)){3}";
var \$ = "[a-fA-F\\\\\\\\d]{1,4}";
var x = \`
(?:
(?:\${\$}:){7}(?:\${\$}|:)|                                    // 1:2:3:4:5:6:7::  1:2:3:4:5:6:7:8
(?:\${\$}:){6}(?:\${c}|:\${\$}|:)|                             // 1:2:3:4:5:6::    1:2:3:4:5:6::8   1:2:3:4:5:6::8  1:2:3:4:5:6::1.2.3.4
(?:\${\$}:){5}(?::\${c}|(?::\${\$}){1,2}|:)|                   // 1:2:3:4:5::      1:2:3:4:5::7:8   1:2:3:4:5::8    1:2:3:4:5::7:1.2.3.4
(?:\${\$}:){4}(?:(?::\${\$}){0,1}:\${c}|(?::\${\$}){1,3}|:)| // 1:2:3:4::        1:2:3:4::6:7:8   1:2:3:4::8      1:2:3:4::6:7:1.2.3.4
(?:\${\$}:){3}(?:(?::\${\$}){0,2}:\${c}|(?::\${\$}){1,4}|:)| // 1:2:3::          1:2:3::5:6:7:8   1:2:3::8        1:2:3::5:6:7:1.2.3.4
(?:\${\$}:){2}(?:(?::\${\$}){0,3}:\${c}|(?::\${\$}){1,5}|:)| // 1:2::            1:2::4:5:6:7:8   1:2::8          1:2::4:5:6:7:1.2.3.4
(?:\${\$}:){1}(?:(?::\${\$}){0,4}:\${c}|(?::\${\$}){1,6}|:)| // 1::              1::3:4:5:6:7:8   1::8            1::3:4:5:6:7:1.2.3.4
(?::(?:(?::\${\$}){0,5}:\${c}|(?::\${\$}){1,7}|:))             // ::2:3:4:5:6:7:8  ::2:3:4:5:6:7:8  ::8             ::1.2.3.4
)(?:%[0-9a-zA-Z]{1,})?                                             // %eth0            %1
\`.replace(/\\\\s*\\\\/\\\\/.*\$/gm, "").replace(/\\
/g, "").trim();
var n = new RegExp(\`(?:^\${c}\$)|(?:^\${x}\$)\`);
var t = new RegExp(\`^\${c}\$\`);
var v = new RegExp(\`^\${x}\$\`);
var a = (e2) => e2 && e2.exact ? n : new RegExp(\`(?:\${d(e2)}\${c}\${d(e2)})|(?:\${d(e2)}\${x}\${d(e2)})\`, "g");
a.v4 = (e2) => e2 && e2.exact ? t : new RegExp(\`\${d(e2)}\${c}\${d(e2)}\`, "g");
a.v6 = (e2) => e2 && e2.exact ? v : new RegExp(\`\${d(e2)}\${x}\${d(e2)}\`, "g");
var r = a;

// lib/bundle_import/index.js
foo_default();
foo4_default();
component_default();
s();
r();
//# sourceMappingURL=index.js.map
'
`;

snapshot[`bundle: prefix > css import 1`] = `
"/* lib/foo.css */
body {
  color: red;
}
body h1 {
  color: #00f;
}

/* packages/mypackage/styles.css */
.mypackage {
  color: red;
}

/* lib/styles.module.css */
.myClass0011c694 {
  color: pink;
}

/* url:https://esm.sh/@radix-ui/colors@0.1.8/blue.css */
:root {
  --blue1: hsl(206, 100%, 99.2%);
  --blue2: hsl(210, 100%, 98.0%);
  --blue3: hsl(209, 100%, 96.5%);
  --blue4: hsl(210, 98.8%, 94.0%);
  --blue5: hsl(209, 95.0%, 90.1%);
  --blue6: hsl(209, 81.2%, 84.5%);
  --blue7: hsl(208, 77.5%, 76.9%);
  --blue8: hsl(206, 81.9%, 65.3%);
  --blue9: hsl(206, 100%, 50.0%);
  --blue10: hsl(208, 100%, 47.3%);
  --blue11: hsl(211, 100%, 43.2%);
  --blue12: hsl(211, 100%, 15.0%);
}

/* url:https://esm.sh/@radix-ui/colors@0.1.8/gray.css */
:root {
  --gray1: hsl(0, 0%, 99.0%);
  --gray2: hsl(0, 0%, 97.3%);
  --gray3: hsl(0, 0%, 95.1%);
  --gray4: hsl(0, 0%, 93.0%);
  --gray5: hsl(0, 0%, 90.9%);
  --gray6: hsl(0, 0%, 88.7%);
  --gray7: hsl(0, 0%, 85.8%);
  --gray8: hsl(0, 0%, 78.0%);
  --gray9: hsl(0, 0%, 56.1%);
  --gray10: hsl(0, 0%, 52.3%);
  --gray11: hsl(0, 0%, 43.5%);
  --gray12: hsl(0, 0%, 9.0%);
}

/* lib/bundle_import/styles.css */
header {
  color: pink;
}
/*# sourceMappingURL=styles.css.map */
"
`;
